File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/utils/GsonUtil.java
package ch.rmy.android.http_shortcuts.utils;

import com.google.gson.ExclusionStrategy;
import com.google.gson.FieldAttributes;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonParser;

import java.io.Reader;

import ch.rmy.android.http_shortcuts.realm.models.Base;
import io.realm.RealmObject;

public class GsonUtil {

    public static String prettyPrint(String jsonString) {
        try {
            JsonParser parser = new JsonParser();
            JsonObject json = parser.parse(jsonString).getAsJsonObject();
            Gson gson = new GsonBuilder().setPrettyPrinting().create();
            return gson.toJson(json);
        } catch (JsonParseException e) {
            return jsonString;
        }
    }

    public static String toJson(RealmObject item) {
        Gson gson = getJsonBuilder().create();
        return gson.toJson(item);
    }

    public static <T extends RealmObject> T fromJson(String json, Class<T> clazz) {
        Gson gson = getJsonBuilder().create();
        return gson.fromJson(json, clazz);
    }

    public static void exportData(Base base, Appendable writer) {
        Gson gson = getJsonBuilder().setPrettyPrinting().create();
        gson.toJson(base, writer);
    }

    public static String exportData(Base base) {
        Gson gson = getJsonBuilder().setPrettyPrinting().create();
        return gson.toJson(base);
    }

    public static Base importData(Reader reader) {
        Gson gson = getJsonBuilder().create();
        return gson.fromJson(reader, Base.class);
    }

    private static GsonBuilder getJsonBuilder() {
        return (new GsonBuilder()).addSerializationExclusionStrategy(new RealmExclusionStrategy());
    }

    private static class RealmExclusionStrategy implements ExclusionStrategy {

        @Override
        public boolean shouldSkipField(FieldAttributes f) {
            return f.getDeclaringClass().equals(RealmObject.class);
        }

        @Override
        public boolean shouldSkipClass(Class<?> clazz) {
            return false;
        }

    }

}

File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/utils/GsonUtil.kt
package ch.rmy.android.http_shortcuts.utils

import ch.rmy.android.http_shortcuts.actions.ActionDTO
import ch.rmy.android.http_shortcuts.realm.models.Base
import com.google.gson.ExclusionStrategy
import com.google.gson.FieldAttributes
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.JsonParser
import com.google.gson.reflect.TypeToken
import io.realm.RealmObject
import java.io.Reader

object GsonUtil {

    fun prettyPrint(jsonString: String): String {
        return try {
            val parser = JsonParser()
            val json = parser.parse(jsonString)
            val gson = GsonBuilder().setPrettyPrinting().create()
            gson.toJson(json)
        } catch (e: JsonParseException) {
            jsonString
        }
    }

    fun toJson(item: RealmObject): String = gson.toJson(item)

    fun <T : RealmObject> fromJson(json: String, clazz: Class<T>): T = gson.fromJson(json, clazz)

    fun exportData(base: Base, writer: Appendable) {
        getPrettyGson().toJson(base, writer)
    }

    fun exportData(base: Base): String = getPrettyGson().toJson(base)

    fun importData(reader: Reader): Base = gson.fromJson(reader, Base::class.java)

    fun <T> fromJsonObject(jsonObject: String?): Map<String, T> {
        if (jsonObject == null) {
            return emptyMap()
        }
        val type = object : TypeToken<Map<String, T>>() {
        }.type
        return gson.fromJson(jsonObject, type)
    }

    fun parseActionList(jsonList: String?): List<ActionDTO> {
        if (jsonList == null) {
            return emptyList()
        }
        val type = object : TypeToken<List<ActionDTO>>() {
        }.type
        return gson.fromJson<List<ActionDTO>>(jsonList, type)
    }

    fun toJson(item: Map<String, Any>): String = gson.toJson(item)

    fun toJson(list: List<Any>): String = gson.toJson(list)

    fun parseString(jsonString: String): JsonElement = JsonParser().parse(jsonString)

    private val gson: Gson by lazy {
        GsonBuilder()
                .addSerializationExclusionStrategy(RealmExclusionStrategy())
                .create()
    }

    private fun getPrettyGson() =
            GsonBuilder()
                    .addSerializationExclusionStrategy(RealmExclusionStrategy())
                    .setPrettyPrinting()
                    .create()

    private class RealmExclusionStrategy : ExclusionStrategy {

        override fun shouldSkipField(f: FieldAttributes) = f.declaringClass == RealmObject::class.java

        override fun shouldSkipClass(clazz: Class<*>) = false

    }

}
